File: backend/app/models/trend.py
from typing import Optional

from pydantic import BaseModel


class KlineBar(BaseModel):
    time: str
    open: float
    high: float
    low: float
    close: float
    volume: float


class KlineData(BaseModel):
    code: str
    kl_type: str
    bars: list[KlineBar]


class IndicatorSet(BaseModel):
    ma_5: list[Optional[float]]
    ma_20: list[Optional[float]]
    ma_60: list[Optional[float]]
    ma_200: list[Optional[float]]
    macd_line: list[Optional[float]]
    macd_signal: list[Optional[float]]
    macd_histogram: list[Optional[float]]
    rsi_14: list[Optional[float]]
    bollinger_upper: list[Optional[float]]
    bollinger_mid: list[Optional[float]]
    bollinger_lower: list[Optional[float]]


class SRLevel(BaseModel):
    level: float
    type: str  # "support" or "resistance"
    strength: int


class SrLevelsResponse(BaseModel):
    supports: list[SRLevel]
    resistances: list[SRLevel]


class StockInfo(BaseModel):
    code: str
    name: str
    market: str

File: frontend/src/types/trend.ts
export interface KlineBar {
  time: string
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface KlineData {
  code: string
  kl_type: string
  bars: KlineBar[]
}

export interface IndicatorSet {
  ma_5: (number | null)[]
  ma_20: (number | null)[]
  ma_60: (number | null)[]
  ma_200: (number | null)[]
  macd_line: (number | null)[]
  macd_signal: (number | null)[]
  macd_histogram: (number | null)[]
  rsi_14: (number | null)[]
  bollinger_upper: (number | null)[]
  bollinger_mid: (number | null)[]
  bollinger_lower: (number | null)[]
}

export interface SRLevel {
  level: number
  type: 'support' | 'resistance'
  strength: number
}

export interface SrLevelsResponse {
  supports: SRLevel[]
  resistances: SRLevel[]
}

export interface StockInfo {
  code: string
  name: string
  market: string
}
